design/fapi-grant-management.md

```diff
@@ -0,0 +1,261 @@
+
+- **scopes**
+
+JSON array where every entry contains the scope parameter value and (optionally) any resource parameter value as defined in [@!RFC8707] passed in the same authorization request. The AS MUST maintain the scope and resource values passed in different authorization requests in separate objects of the JSON structure in order to preserve their relationship.
+
+- **claims**
+
+JSON array containing the names of all OpenID Connect claims (see [@!OpenID]) as requested and consented in one or more authorization requests associated with the respective grant.
+
+- **authorization_details**
+
+JSON Object as defined in [@!I-D.ietf-oauth-rar] containing all authorization details as requested and consented in one or more authorization requests associated with the respective grant.
+
+- **status**
+
+- **client_id**
+
+- **created_at**
+
+- **updated_at**
+
+
+
+
+### OAuth Protocol Extensions
+#### Authorization Request
+The spec introduces the authorization request parameters `grant_id` and `grant_management_action`. These parameters can be used with any request serving as authorization request.
+
+- `grant_id`: OPTIONAL. String value identifying an individual grant managed by a particular authorization server for a certain client and a certain resource owner.
+
+- `grant_management_action`: string value controlling the way the authorization server shall handle the grant when processing an authorization request. The spec defines the following values:
+
+ `create`: the AS will create a fresh grant if the AS supports the grant management action create.
+
+ `update`: this mode requires the client to specify a grant id using the grant_id parameter. If the parameter is present and the AS supports the grant management action update, the AS will assign all permissions as consented by the user in the actual request to the respective grant.
+The following example shows how a client may ask the authorization request to use a certain grant id:
+
+````
+GET /authorize?response_type=code&
+     client_id=s6BhdRkqt3
+     &grant_management_action=update
+     &grant_id=TSdqirmAxDa0_-DB_1bASQ
+     &scope=write
+     &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
+     &code_challenge_method=S256
+     &code_challenge=K2-ltc83acc4h... HTTP/1.1
+Host: as.example.com 
+````
+#### Authorization Error Response
+In case the `grant_id` is unknown or invalid, the authorization server will respond with an error code `invalid_grant_id`.
+
+in case the AS does not support a grant management action requested by the client, it will respond with the error code `invalid_request`.
+
+#### Token Response
+The spec introduces the token response parameter `grant_id`:
+
+`grant_id`: URL safe string value identifying an individual grant managed by the AS for a certain client and a certain resource owner. The grant_id value MUST be unique in the context of a certain authorization server and SHOULD have enough entropy to make it impractical to guess it.
```

Comment:
How many entropy does the generation of `grant_id`? If you've already determined how to generate `grant_id`, it might be good to write it and the expected entropy this generator has, and remove `SHOULD have enough entropy to make it impractical to guess it.`.

Reply:
we will use [https://tools.ietf.org/html/rfc4122](urhttps://tools.ietf.org/html/rfc4122l) spec to generate the grant_id.

Reply:
I see. Thank you for your clarification.